### PR TITLE
fix: do not send context updates on context read

### PR DESCRIPTION
**Pull request by ganeshnj**
### What and why?

TLDR: when reading `context` from `FeatureScope` triggers a context update which again publishes context updates again to all the subscribes and we end up in a loop.

This shoots up CPU usage and an app termination likely.

### How?

- Now, when we read the context, we no longer update it.
- a small improvement in WT message receiver

There a few learnings from this bug
- The current pub-sub mechanism is very complex and requires a lot of dive deep before you can start making sense.
- The value pub-sub delivers to its consumers is very simple which we have talked before also. For example, on every context, telemetry, webview or baggage message, each receiver is notified which is wastage of CPU cycles. The receivers must be able to subscribe to individual message or part of message.

### Review checklist
- [ ] Feature or bugfix MUST have appropriate tests (unit, integration)
- [ ] Make sure each commit and the PR mention the Issue number or JIRA reference
- [ ] Add CHANGELOG entry for user facing changes

### CODE REVIEW

File: DatadogCore/Sources/Core/Context/DatadogContextProvider.swift [hunk 1]
```diff
@@ -110,7 +108,12 @@ internal final class DatadogContextProvider {
     ///
     /// - Parameter block: The block closure called with the current context.
     func write(block: @escaping (inout DatadogContext) -> Void) {
-        queue.async { block(&self.context) }
+        queue.async {
```
Comment (mariedm) [inline DatadogCore/Sources/Core/Context/DatadogContextProvider.swift:111]: Given what you mentioned in the PR description, should we check if the context changed before notifying the receivers?

**Review by mariedm — APPROVED** [review]

File: DatadogCore/Sources/Core/Context/DatadogContextProvider.swift (hunk 1 above)
Comment (ganeshnj) [reply DatadogCore/Sources/Core/Context/DatadogContextProvider.swift:111]: this is not so possible with current context definition, we are going to do some followup with a larger refactor.

**Review by ncreated — APPROVED** [review]
🎯 Well done 👍

> The current pub-sub mechanism is very complex and requires a lot of dive deep before you can start making sense.
The value pub-sub delivers to its consumers is very simple which we have talked before also. For example, on every context, telemetry, webview or baggage message, each receiver is notified which is wastage of CPU cycles. The receivers must be able to subscribe to individual message or part of message.

💯 agree, looking forward to follow up on this
